### ISSUE TITLE
Search then Ask on Documents returned from Search?

### ISSUE THREAD

**Issue by JohnGalt1717** [microsoft/kernel-memory#198]
I have some documents indexed into Qdrant.

What I'm trying to do is limit the possible documents for it to synthesyse a response from so that the response is faster, and obviously context length becomes less of an issue.

When I run this:

`var searchResults = await memory.SearchAsync(request.Question, index, cancellationToken: cancellationToken);`

I get a series of citations back.

But the citations don't have a documentId

So I have no way of adding filters.

I tried to hack it like this:  

```
		var documentIds = searchResults.Results
						.Select(t => t.Link[..t.Link.IndexOf('/')])
						.Distinct()
						.ToArray();

                  var answer = await memory.AskAsync(request.Question, index, filters: documentIds.Select(r => MemoryFilters.ByTag("__document_id", r)).ToArray(), cancellationToken: cancellationToken);

```
But I just get "no results"

So my question is 2 fold:

Can the citations have the documentId included please as a property so that this is easay to correlate to a database? (we use this for security so that we strip all documents that the user doesn't have access to based on information in the database)

And in the mean time, is there any way to get AskAsync to use something from the search results to be able to Ask?

Ultimately, it would be great if this was part of a standard process like you can do with langChain where it does a search first, then answers the questions on the partition values returned iterating through, pulling the possible info, and then answering.  This will fix the context size entirely.

**Comment by dluc** [comment]
Thanks for raising this request, this PR should provide the details you're looking for: https://github.com/microsoft/kernel-memory/pull/201

> And in the mean time, is there any way to get AskAsync to use something from the search results to be able to Ask?

Ask allows to filter by index, document ID and tags, which should cover your scenario. The code searches for relevant chunks and uses only those to generate an answer.

could you provide an example where this approach doesn't work?

> Ultimately, it would be great if this was part of a standard process like you can do with langChain where it does a search first, then answers the questions on the partition values returned iterating through, pulling the possible info, and then answering. This will fix the context size entirely.

If I understand this correctly, it would be more expensive right? e.g. the same prompt would be executed more times. We plan to provide more strategies, so perhaps we'll add this as one of them, allowing the caller to select which strategy to adopt.

**Comment by JohnGalt1717** [comment]
Great! #201 looks amazing!

With #201 I think that the Ask filters will work fine.

My assumption is that the Search (and it appears to be true because it doesn't take forever) is doing a vector cosine similarity index lookup.  My assumption is that this would then limit the results that the AskAsync function does.  I looked in the source and it appears that AskAsync is already doing this so this probably isn't relevant and can be ignored.

**Comment by dluc** [comment]
FYI #201 merged, version 0.21 should be out in few minutes.

New fields in the response:

* `index`: name of the memory index (aka collection in the vector store) used to retrieve relevant sources
* `documentId`: ID of the document(s) used. This can be useful to fetch/update/delete etc.
* `fileId`: for those cases where a document contains multiple files (e.g. a text file plus some images), ID of the file used. This value can be used to filter further, see the tags in the JSON response.
* `noResult`: a boolean indicating whether the Search returned no results, and Ask returned no answer. This allows to quickly check the response without having to count or parse the response content.
* `noResultReason`: a brief explanation about why there was no response (when `noResult` is `true`)
* `sourceUrl`: used only when the source is a web page, with the URL provided during the import

**Comment by dluc** [comment]
One example:
```json
{
  "question": "what is ABCD?",
  "noResult": true,
  "noResultReason": "No relevant memories found",
  "text": "INFO NOT FOUND.",
  "relevantSources": [
    {
      "link": "default/img001/ff8817ad8b41445f9698b4dcc0677623",
      "index": "default",
      "documentId": "img001",
      "fileId": "ff8817ad8b41445f9698b4dcc0677623",
      "sourceContentType": "image/jpeg",
      "sourceName": "file6-ANWC-image.jpg",
      "partitions": [
        {
          "text": "August 28, 2023\nf\nin\nAI\nloT\nMicrosoft is a proud platinum sponsor of the Automotive News World Congress 2023 event taking place in Detroit, Michigan on September 12, 2023. With a focus on the global automotive industry at large, this event attracts original equipment manufacturers (OEMs), suppliers, and other value chain companies in the broader automotive and mobility sectors.\nThis year's congress event will highlight topics such as autonomy and sustainability with a focus on the reduction of carbon emissions, retail, and the role of governments in developing industrial decision-making. Microsoft will provide thought leadership around our role in supporting transformation in the automotive industry.\nOur focus encapsulates the entire segment known as \"mobility\" with a distinct concentration in automotive, mobility, transportation, and logistics, or effectively, the movement of people and goods. Within automotive, we provide the \"platform for platform creators\" and empower companies to innovate around industry opportunities.\nAccelerating the future of mobility\nTransform to smart mobility.\nDiscover more >",
          "relevance": 0.6985194,
          "lastUpdate": "2023-12-14T02:06:37-08:00",
          "tags": {
            "__document_id": [
              "img001"
            ],
            "__file_type": [
              "image/jpeg"
            ],
            "__file_id": [
              "ff8817ad8b41445f9698b4dcc0677623"
            ],
            "__file_part": [
              "2ddee1693f614f7d8afc05d9a73bede0"
            ]
          }
        }
      ]
    },
...
```

**Comment by JohnGalt1717** [comment]
Can this be made backwards compatible with existing indexes? As it is right now the url (at least) property causes a crash on this line in SearchClient:

string? sourceUrl = memory.Payload[Constants.ReservedPayloadUrlField].ToString();

Looks like the rest work fine and since that's a nullable string anyhow, using TryGetValue would be preferable anyhow.

**Comment by dluc** [comment]
> Can this be made backwards compatible with existing indexes? As it is right now the url (at least) property causes a crash

Fixed in nuget version 0.23.231219.1
